contracts/adapters/swap/astroport/src/contract.rs

```diff
@@ -174,19 +160,34 @@ fn execute_swap(
     // Create a response object to return
     let mut response: Response = Response::new().add_attribute("action", "execute_swap");
 
-    // Add an astroport pool swap message to the response for each swap operation
-    for operation in &operations {
-        let swap_msg = WasmMsg::Execute {
-            contract_addr: env.contract.address.to_string(),
-            msg: to_json_binary(&ExecuteMsg::AstroportPoolSwap {
-                operation: operation.clone(),
-            })?,
-            funds: vec![],
-        };
-        response = response.add_message(swap_msg);
+    // Add an astroport pool swap message to the response for each swap operation in each route
+    for route in &routes {
+        for (idx, operation) in route.operations.iter().enumerate() {
+            // if first operation in a route, set offer asset to the route's offer asset
+            // otherwise, the offer asset will be determined by the contracts balance
+            let offer_asset = if idx == 0 {
+                Some(route.offer_asset.clone())
+            } else {
+                None
+            };
+
+            let swap_msg = WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_json_binary(&ExecuteMsg::AstroportPoolSwap {
+                    operation: operation.clone(),
+                    offer_asset: offer_asset,
+                })?,
+                funds: vec![],
+            };
+
+            response = response.add_message(swap_msg);
+        }
     }
 
-    let return_denom = match operations.last() {
+    let return_denom = match routes
+        .last()
+        .and_then(|last_route| last_route.operations.last())
+    {
         Some(last_op) => last_op.denom_out.clone(),
```

Comment:
Yeah let's make sure we do validation on entrypoint that the first op denom / offer asset of all routes must be the same, and the last op denom of all routes must be the same in the entrypoint. Then this line is safe